### DISCUSSION TITLE
Azure Multi-Tenant SSO / Custom Authentication Options / Password Complexity Options

### DISCUSSION THREAD

**Discussion by crsnchng** [Budibase/budibase#5550] [Help]
Hello!

I'm under a bit of pressure to implement Azure Multi-Tenant SSO. Wondering if there was a plan to implement Azure Multi-Tenant SSO support? If not, is there documentation on how to add my own custom authentication option? I have identified where in the existing authentication code this would be added and I can see why multi-tenant doesn't work as is. So, if I can add a custom option, that works as well. 

As a side note, I have noticed that there aren't any native password complexity requirements for the local Budibase users. Is there somewhere I can add complexity requirements/is this planned for a future release?

Thanks in advance.

**Comment by mjashanks** [comment]
Hi @crsnchng 

I am not familiar with Azure Multi-tenant SSO, so forgive my lack of understanding.

The ideal scenario would be for us to modify the existing OpenID connector to support Azure Multi-Tenant. But maybe this is not supported by OpenID - therefor that wouldn't work.

This page lists OpenID, under the "Multi-tenant applications" category: https://docs.microsoft.com/en-us/azure/architecture/multitenant-identity/authenticate - which would suggest that you should be able to use OpenID.

> I can see why multi-tenant doesn't work as is

Can you link to this in the code for us?

RE password complexity - there is no way to control this anywhere in Budibase. Our assumption here is that people should use SSO for this! Of course, this falls over if you are using a mix of SSO and basic auth. Is this the case for you?

**Reply by crsnchng** [reply]
Hi @mjashanks,

No worries, thank you for the reply.

It does work with OIDC - in fact, as it currently is implemented, we get about 90% of the way there. You will just need to modify the OIDC code to accept multiple issuer values: https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant#update-your-code-to-handle-multiple-issuer-values

When the application on Azure is changed from single tenant to multi tenant, the config URL changes to: https://login.microsoftonline.com/organizations/v2.0/.well-known/openid-configuration and there is currently nowhere in the Budibase auth code to take this and handle multiple issuer values. The link I sent to the Microsoft documentation explains everything in better detail.

In terms of the password complexity, ideally we can get the Azure multi tenant auth implemented which would remove my need for basic auth. 

It seems like it could be a pretty quick addition to the existing OIDC code - please let me know what you find and if you can implement it. I cannot fully implement my Budibase app without it.

Thank you!

**Reply by crsnchng** [reply]
@mjashanks - apologies for the double reply - The code I was looking at was here: packages/backend-core/src/middleware/passport/oidc.js

I see you are using passport - was able to find this, it seems they have a strategy pre-written: http://www.passportjs.org/packages/passport-azure-ad/

**Reply by crsnchng** [reply]
Hi @mjashanks, hope all is well. Just wanted to bump this to see if there were any changes being queued up as a result of this thread or what the status may be. 

I appreciate any and all assistance with this.

Best,

C

**Reply by aderici** [reply]
I guess the best you could do is to fork the repo and make your own changes? 

Just to add some summary as i've been there before, you can use microsoft login to allow "only internal employees" to login or you can use it to allow "anyone with a microsoft work account" to login. If you chose the latter, you have to disable "issuer verification" in openid id token verification logic.

But there seems to be a bigger problem though. How are you matching a microsoft user to an internal user? I believe through "email" claim. But "email" field in microsoft world is a completely unverified free field so a user can present themselves with any unverified custom email. In a multi-tenant microsoft login implementation that can create serious security problems if you don't isolate tenants. 

The correct claim to check is called "upn" which is in an email format, but microsoft won't allow you to put any domain that your company hasnt verified ownership so you cant present yourself as mark@facebook

I believe budibase isn't meant for public facing multi-tenant apps yet,

**Reply by crsnchng** [reply]
@aderici Totally understand all of the points you make - thanks for the reply. I will look into other methods or potentially fork the repo. 

Currently, I'm using the native Budibase code which I believe uses the email claim, as you mention. Ideally, I only allow access from the tenants that should have access. I absolutely understand that Budibase isn't meant for public facing apps. The multi tenant aspect of my Budibase implementation isn't intended to be public. Ideally I would leverage the correct security features from Azure and OIDC to limit access to permitted tenants.

Best,

C
